--- wine-1.9.17/dlls/ntdll/sync.c.orig	2016-09-11 19:58:02.411397790 -0400
+++ wine-1.9.17/dlls/ntdll/sync.c	2016-09-11 19:58:51.855663444 -0400
@@ -734,11 +778,6 @@
 
     if (!len && use_pipe_objects())
     {
-        req->access  = access;
-        req->owned   = InitialOwner;
-        wine_server_add_data( req, objattr, len );
-        status = wine_server_call( req );
-        *MutantHandle = wine_server_ptr_handle( reply->handle );
         status = create_pipe_object(MutantHandle, MUTEX_MAGIC, !InitialOwner);
     }
     else
