--- wine-l-pa-1.7.8/dlls/gdiplus/graphics.c.orig	2013-12-16 23:47:13.867529956 -0500
+++ wine-l-pa-1.7.8/dlls/gdiplus/graphics.c	2013-12-16 23:47:01.622530686 -0500
@@ -3371,7 +3371,12 @@
 
     if (!graphics->hdc)
     {
-        FIXME("graphics object has no HDC\n");
+        static int quietfixme = 0;
+        if (quietfixme == 0)
+        {
+            FIXME("graphics object has no HDC\n");
+            quietfixme = 1;
+        }
         return Ok;
     }
 
@@ -5709,7 +5714,12 @@
 
     if (!graphics->hdc)
     {
-        FIXME("graphics object has no HDC\n");
+        static int quietfixme = 0;
+        if (quietfixme == 0)
+        {
+            FIXME("graphics object has no HDC\n");
+            quietfixme = 1;
+        }
         return Ok;
     }
 
